Sources/Build/misc.swift

```diff
@@ -29,8 +32,102 @@ func platformFrameworksPath() throws -> String {
 }
 
 extension CModule {
+    
+    var moduleMap: String {
+        return "module.modulemap"
+    }
+    
     var moduleMapPath: String {
-        return Path.join(path, "module.modulemap")
+        return Path.join(path, moduleMap)
+    }
+}
+
+extension ClangModule {
+    
+    public enum ModuleMapError: ErrorProtocol {
+        case UnsupportedIncludeLayoutForModule(String)
+    }
+    
+    public func generateModuleMap(inDir wd: String) throws {
+        
+        ///Return if module map is already present
+        guard !moduleMapPath.isFile else {
+            return
+        }
+        
+        let includeDir = path
+        
+        ///Warn and return if no include directory
+        guard includeDir.isDirectory else {
+            print("warning: No include directory found, a library can not be imported without any public headers.")
+            return
+        }
+        
+        let walked = walk(includeDir, recursively: false).map{$0}
+        
+        let files = walked.filter{$0.isFile && $0.hasSuffix(".h")}
+        let dirs = walked.filter{$0.isDirectory}
+        
+        ///There are three possible cases for which we will generate modulemaps:
+        ///Flat includes dir with only header files, in that case we generate
+        /// `umbrella "path/to/includes/"`
```

Comment:
I think we should generate `umbrella header "foo.h"` even for the flat layout, if "foo.h" exists.